### PR TITLE
fix(runtime): rewind GDN state on session prefix-cache hits

### PR DESCRIPTION
**Pull request by ewindisch**
Found by static recon, confirmed with a failing test: the session prefix-cache path called `truncate_to(prefix_len)` on the full-attention KV but never rewound Qwen3.5 GDN/DeltaNet `conv_states`/`rec_states`, which are single-slot and model-global. On a multi-turn prefix hit, decode resumed with recurrent state from the *previous turn's generation* — silently wrong context for hybrid models. (Prefix miss and Stateless requests were already safe: `clear_kv_cache` resets SSM state.)

## Fix
- `KVCacheManager` stores a per-session SSM snapshot, captured at end of prefill and saved alongside `cached_token_ids` (one per session, follows the session cache lifecycle)
- The prefix-hit path restores it via `rewind_session_state`, reusing the existing TTT snapshot/restore mechanism (`snapshot_ssm_states`/`restore_ssm_states`, deep-copied)

## Regression tests (4, in qwen3_5 pipeline_tests)
- `session_prefix_hit_restores_ssm_state_matches_fresh_cache` — two-turn session, shared prefix, output must match a fresh-cache reference token-for-token
- `session_prefix_hit_without_ssm_rewind_diverges_from_fresh_cache` — divergence demo documenting the old behavior
- `session_partial_prefix_match_forces_full_recompute`, `clear_kv_cache_fully_resets_ssm_state`

## Validation (BuildQ helper, rocm71, CPU)
- `cargo test -p hyprstream --lib --tests --no-fail-fast` — 1640 passed / 0 failed / 3 ignored (auth::pglite_store skipped: pre-existing embedded-postgres harness crash on this host, reproduces even with --test-threads=1 here; unrelated to this diff — worth a look from the pglite lane owner, two lanes have now hit it)
- `cargo clippy -p hyprstream --all-targets -- -D warnings` — clean

docs/KV-CACHE-ARCHITECTURE.md updated for the hybrid prefix-caching behavior.

### CODE REVIEW

File: crates/hyprstream/src/runtime/kv_cache.rs [hunk 1]
```diff
@@ -2168,8 +2180,40 @@ impl KVCacheManager {
...
+    /// The snapshot is what makes a later prefix hit rewindable for recurrent
+    /// layers: KV truncates to the matched prefix, the snapshot restores the
+    /// GDN conv/rec state to the end of the cached sequence.
+    pub fn set_cached_tokens_with_ssm(
+        &mut self,
+        tokens: Vec<i64>,
+        ssm_snapshot: Option<(Vec<Option<Tensor>>, Vec<Option<Tensor>>)>,
+    ) {
         self.cached_token_ids = tokens;
+        self.ssm_snapshot = ssm_snapshot;
```
[the remark this answers is not in the document]
Comment (ewindisch) [reply crates/hyprstream/src/runtime/kv_cache.rs:2257]: Addressed by commit 5ce39cc04 (carried to the current head 744c5a52d1baa0cb71ee79e3c1b2ef677f1342ac through merge 1cde7f783, which resolved the qwen3_5.rs test-tail conflict by keeping both this PR's session tests and main's causal-mask test).

KVCacheManager::memory_usage() now sums the snapshot tensors' allocated bytes (numel × dtype_element_size per conv/rec slot) on top of the layer caches, so evict_to_budget's accounting includes the snapshot VRAM and cannot skip offloading while snapshots consume untracked budget.

Regression: test_memory_usage_includes_ssm_snapshot — usage grows by exactly the snapshot's byte count when a snapshot is recorded and shrinks back when a token-only recording drops it.

Current exact head: 744c5a52d1baa0cb71ee79e3c1b2ef677f1342ac (perf/gdn-prefix-state).

File: crates/hyprstream/src/runtime/torch_engine.rs [hunk 2]
```diff
@@ -1604,6 +1604,58 @@ impl TorchEngine {
...
+            cache_guard.ssm_snapshot()
+        } else {
+            None
+        }
+    };
+    match snapshot {
+        Some((conv_snap, rec_snap)) => {
+            cache.lock().truncate_to(prefix_len);
+            q35.restore_ssm_states(conv_snap, rec_snap);
+            prefix_len
```
[the remark this answers is not in the document]
Comment (ewindisch) [reply crates/hyprstream/src/runtime/torch_engine.rs:1651]: Fixed by commit 614ecb3f1 (at current head 744c5a52d1baa0cb71ee79e3c1b2ef677f1342ac).

The reuse decision now lives in resolve_session_prefill_start (torch_engine.rs): only a strict partial hit (0 < prefix_len < prompt_len) rewinds via rewind_session_state. An exact-prompt repeat (prefix_len == cached_token_count() == prompt_len) clears the cache — clear_kv_cache resets both KV and the GDN conv/rec state — and recomputes. The cache does not retain last-position logits, so this is the recompute-from-cleared-state branch of your recommendation, not the retained-logits one.

Tests (qwen3_5 pipeline_tests):
- session_exact_prompt_repeat_forces_cleared_recompute — resolve must return 0 with fully cleared conv/rec state, and the recomputed turn must match a fresh-cache reference token-for-token and logit-for-logit (1e-5). Key assertions are exact, so it is non-vacuous by construction.
- session_exact_prompt_hit_restored_reforward_diverges_from_fresh_cache — characterization: restore-then-reforward (the old behavior) diverges from a fresh prefill at the FIRST re-forwarded row (asserted >1e-4; measured ~6.9e-4; the divergence decays row-wise as the delta-rule forgets, which is why last-row-only observation under-sells it).

Current exact head: 744c5a52d1baa0cb71ee79e3c1b2ef677f1342ac (perf/gdn-prefix-state).

File: crates/hyprstream/src/runtime/kv_cache.rs [hunk 3]
```diff
@@ -2223,6 +2289,11 @@ impl KVCacheManager {
         for mut cache_ref in self.layer_caches.iter_mut() {
             cache_ref.to_device(device);
         }
+        if let Some((conv, rec)) = &mut self.ssm_snapshot {
+            for t in conv.iter_mut().chain(rec.iter_mut()).flatten() {
+                *t = t.to_device(device);
```
[the remark this answers is not in the document]
Comment (ewindisch) [reply crates/hyprstream/src/runtime/kv_cache.rs]: Fixed by commit 744c5a52d (current head 744c5a52d1baa0cb71ee79e3c1b2ef677f1342ac).

KVCacheManager now records each snapshot slot's owning device from its own tensor when the snapshot is stored (ssm_slot_devices, one entry per layer). restore_to_gpu sends each conv/rec slot back to its recorded owning device instead of the single call argument, falling back to the argument only when a snapshot predates recording — so get_or_create's single cuda_if_available() restore no longer drags layer-1+ recurrent state onto CUDA 0 under a multi-GPU LayerDeviceMap, and the next GDN forward sees each state on its own layer's device. offload_to_cpu keeps the record: offloading moves slots to CPU but does not change ownership.

Regression: test_restore_moves_snapshot_slots_to_owning_devices builds the exact hazard shape — mixed Cuda(0)/Cpu owning devices across conv/rec slots, offload_to_cpu, then restore with the pre-fix call shape (cuda_if_available) — and asserts each slot returns to its owning device with values intact. Non-vacuity check: reverting the restore loop to the single-device form fails this test with 'conv slot 1 must return to its owning device' (ran on an RTX 5090 host).

Current exact head: 744c5a52d1baa0cb71ee79e3c1b2ef677f1342ac (perf/gdn-prefix-state).

File: crates/hyprstream/src/runtime/torch_engine.rs [hunk 4]
```diff
@@ -1605,6 +1605,91 @@ impl TorchEngine {
...
+/// Returns the prefill start position (0 = cache cleared, full prefill).
+pub(crate) fn resolve_session_prefill_start(
+    model: &dyn ModelOperations,
+    prefix_len: usize,
+    prompt_len: usize,
+) -> usize {
+    if prefix_len > 0 && prefix_len < prompt_len {
+        rewind_session_state(model, prefix_len)
+    } else {
+        model.clear_kv_cache();
```
[the remark this answers is not in the document]
Comment (ewindisch) [reply crates/hyprstream/src/runtime/torch_engine.rs:1688]: Fixed by commit 6c3b9284a (current exact head 6c3b9284ab9c3408d30d8d605df76b0ac0b55a29 on perf/gdn-prefix-state).

Implemented the fail-safe recording gate rather than deferring the clear, preserving the exact-prompt semantics and the hybrid rewind path untouched: save_cached_tokens now returns early unless prefill_time_ms is Some — and sample_next_token sets it exactly when the first prefill forward returned successfully (the ? on the forward propagates before the assignment). A drop before the first poll (including the drive_generation_loop biased-cancellation path you identified), or a failed first prefill, therefore records nothing; the next turn sees an empty cache and misses cleanly instead of claiming a prefix over missing/zero KV entries. The constructor-time clear itself stays: it is fail-safe for hybrid models (no snapshot ⇒ forced recompute) and the fresh-prefill path restores state before any recording happens.

Regression: drop_before_first_poll_does_not_record_session_cache (torch_engine tests) drives a real TextStream against a stub ModelOperations serving a real, empty KVCacheManager, drops the stream before the first poll, and asserts (1) cached_token_count() == 0 and (2) prefix_match_len over an extending prompt == 0 — no prefix claimed over state that was never computed. Non-vacuity: reverting the guard fails the test with recorded count 2 vs 0 (the prompt length), i.e. exactly the unguarded-recording bug.

Focused gates via BuildQ: torch_engine tests 23/23, kv_cache 26/26, qwen3_5 pipeline 14/14 (exact-prompt and owning-device tests retained), release clippy -D warnings clean, full lib suite green except the 31 known pre-existing env-flaky auth::key_rotation composite_* members (skipped: 1841 passed / 0 failed / 1 ignored).

Current exact head: 6c3b9284ab9c3408d30d8d605df76b0ac0b55a29.

File: crates/hyprstream/src/runtime/torch_engine.rs [hunk 5]
```diff
@@ -3584,7 +3794,24 @@ impl<'a> TextStream<'a> {
...
+        // Drop with the cache already cleared — exact-hit/miss resolution runs
+        // at generation setup, before any forward — and a failed first prefill
+        // can leave partial forward state matching no token count. Recording
+        // prompt IDs in either case lets a later extending prompt claim a
+        // prefix_len over missing or wrong state (a pure-attention partial
+        // prefill would then resume with no history behind it). `prefill_time_ms`
+        // is set exactly when the first prefill forward returned successfully
+        // (see `sample_next_token`), so it is the fail-safe recording gate.
+        if self.prefill_time_ms.is_none() {
+            return;
```
[the remark this answers is not in the document]
Comment (ewindisch) [reply crates/hyprstream/src/runtime/torch_engine.rs:3907]: Fixed by commit 16a1b8d9 (current exact head 16a1b8d92e24d37bb7531023c00a1357099e1c45 on perf/gdn-prefix-state).

The invalidation now happens at the boundary itself: KVCacheManager::clear_all drops cached_token_ids, the SSM snapshot, and the slot-device record together with the layer tensors, so every clear path — the exact-hit/miss resolution via model.clear_kv_cache(), the engine's no-match clear, and the tenant/compat stale wipes — can no longer leave describable metadata over emptied KV. (llama's clear_kv_cache binding gains the matching mut; the set-up semantics for hybrid models, exact-prompt recompute, owning-device restore, and the cancellation recording guard are unchanged.)

Regressions, both non-vacuous:
- clear_all_drops_cached_token_and_ssm_metadata (kv_cache tests) pins the manager-level contract: IDs, snapshot, and slot-device record all gone after clear_all, and an extending prompt claims no prefix.
- setup_clear_drops_prior_session_metadata_on_cancelled_turn (torch_engine tests) drives the full production chain you outlined: a compatibly stamped prior session cache holding prompt IDs + end-of-prefill SSM snapshot → an exact-hit turn whose set-up clear routes through resolve_session_prefill_start → model.clear_kv_cache() → drop before the first poll → an extending prompt sees no prefix, no stale snapshot, no token count. Reverting the metadata wipes fails both tests with the prior count retained (3 vs 0).

Focused gates via BuildQ: torch_engine tests 24/24, kv_cache 27/27, qwen3_5 pipeline 14/14 (all prior cancellation/exact-prompt/owning-device tests retained and green), release clippy -D warnings clean, full lib suite excluding the 31 known env-flaky composite_* members: 1843 passed / 0 failed / 1 ignored.

Current exact head: 16a1b8d92e24d37bb7531023c00a1357099e1c45.

**Comment by ewindisch** [comment]
Backlog flattening update: this remains an active correctness fix, but its current hosted run is blocked only by the repository-wide loopback-burn-down check detecting three main-carried literals in . No unrelated baseline change will be added to this PR. Reconcile against current main after the loopback policy owner resolves that baseline, then obtain fresh review/checks and merge or close.

**Comment by ewindisch** [comment]
Correction to the backlog update: this remains an active correctness fix, but its current hosted run is blocked only by the repository-wide loopback-burn-down check detecting three main-carried literals in `crates/hyprstream-pds/src/repo_authority.rs`. No unrelated baseline change will be added to this PR. Reconcile against current main after the loopback policy owner resolves that baseline, then obtain fresh review/checks and merge or close.
